sklearn/cluster/optics_.py

```diff
@@ -527,3 +565,355 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+
+    Parameters
+    ----------
+    steep_point : bool array, shape (n_samples)
+        True if the point is steep downward (upward).
+
+    xward_point : bool array, shape (n_samples)
+        True if the point is an upward (respectively downward) point.
+
+    start : integer
+        The start of the xward region.
+
+    min_samples : integer
+       The same as the min_samples given to OPTICS. Up and down steep
+       regions can't have more then ``min_samples`` consecutive non-steep
+       points.
+
+    Returns
+    -------
+    index : integer
+        The current index iterating over all the samples, i.e. where we are up
+        to in our search.
+
+    end : integer
+        The end of the region, which can be behind the index. The region
+        includes the ``end`` index.
+    """
+    n_samples = len(steep_point)
+    non_xward_points = 0
+    index = start
+    end = start
+    # find a maximal area
+    while index < n_samples:
+        if steep_point[index]:
+            non_xward_points = 0
+            end = index
+        elif not xward_point[index]:
+            # it's not a steep point, but still goes up.
+            non_xward_points += 1
+            # region should include no more than min_samples consecutive
+            # non steep xward points.
+            if non_xward_points > min_samples:
+                break
+        else:
+            return min(index, n_samples - 1), end
+
+        index += 1
+    return min(index + 1, n_samples - 1), end
+
+
+def _extend_upward_region(steep_upward_mask, downward_mask, start,
+                          min_samples):
+    return _extend_region(steep_upward_mask, downward_mask,
+                          start, min_samples)
+
+
+def _extend_downward_region(steep_downward_mask, upward_mask, start,
+                            min_samples):
+    return _extend_region(steep_downward_mask, upward_mask,
+                          start, min_samples)
+
+
+def _update_filter_sdas(sdas, mib, xi_complement, reachability_plot):
+    """Update steep down areas (SDAs) using the new maximum in between (mib)
+    value, and the given complement of xi, i.e. ``1 - xi``.
+    """
+    if np.isinf(mib):
+        return []
+    res = [sda for sda in sdas
+           if mib <= reachability_plot[sda['start']] * xi_complement]
+    for sda in res:
+        sda['mib'] = max(sda['mib'], mib)
+    return res
+
+
+def _correct_predecessor(reachability_plot, predecessor, s, e):
+    """Correct for predecessors.
+
+    Applies Algorithm 2 of [1]_.
+
+    Input parameters are ordered by the computer OPTICS ordering.
+
+    .. [1] Schubert, Erich, Michael Gertz.
+       "Improving the Cluster Structure Extracted from OPTICS Plots." Proc. of
+       the Conference "Lernen, Wissen, Daten, Analysen" (LWDA) (2018): 318-329.
+    """
+
+    while s < e:
+        if reachability_plot[s] > reachability_plot[e]:
+            return s, e
+        for i in range(s, e):
+            if predecessor[e] == i:
```

Comment:
This is wrong. This should be
```if predecessor[e] == reachability_plot[i]```
and it supposedly is faster to access `predecessor[e]` only once.

Reply:
Thanks @kno10 , this fixed our issue. Ping @assiaben